src/nvapi_gpu.cpp

```diff
@@ -169,4 +170,139 @@ extern "C" {
 
         return Ok(n);
     }
+
+    NvAPI_Status __cdecl NvAPI_GPU_GetDynamicPstatesInfoEx(NvPhysicalGpuHandle hPhysicalGpu, NV_GPU_DYNAMIC_PSTATES_INFO_EX *pDynamicPstatesInfoEx) {
+        constexpr auto n = "NvAPI_GPU_GetDynamicPstatesInfoEx";
+        static bool alreadyLoggedNoNvml = false;
+        static bool alreadyLoggedOk = false;
+
+        if (nvapiAdapterRegistry == nullptr)
+            return ApiNotInitialized(n);
+
+        if (pDynamicPstatesInfoEx == nullptr)
+            return InvalidArgument(n);
+
+        if (pDynamicPstatesInfoEx->version != NV_GPU_DYNAMIC_PSTATES_INFO_EX_VER)
+            return IncompatibleStructVersion(n);
+
+        auto adapter = reinterpret_cast<NvapiAdapter*>(hPhysicalGpu);
+        if (!nvapiAdapterRegistry->IsAdapter(adapter))
+            return ExpectedPhysicalGpuHandle(n);
+
+        if (!adapter->HasNvml())
+            return NoImplementation(str::format(n, ": NVML not loaded"), alreadyLoggedNoNvml);
```

Comment:
This error can be misleading: because `adapter->HasNvml()` can be false even if NVML was loaded and initialized successfully (due to missing `nvmlDevice_t` pointer). I guess in that case we should probably return `HANDLE_INVALIDATED` but I think it would make more sense to ignore non–Nvidia GPUs instead.

Reply:
I agree that having more fine grained return codes would be nice. In reality though, I think most consumer only ever check that the return code is OK/not OK, but don't contain code paths for specific error codes. Fine grained error code are important when developing against an API, though I hope that nobody is developing against our nvapi implemenation. So I'm actually Ok here to return `NoImplemenation` for both nvml not available (no wine-nvml) or when something went wrong when loading nvml (e.g. AMD GPU or sandboxed environments like Proton). If this bites us, we can always rethink this topic.

Reply:
It's not always technically loading/initializing NVML: a scenario where the user has more than one GPU, one from Nvidia and one from another vendor, is not that uncommon. In that case NVML will successfully load and find `nvmlDevice_t` for NV GPU but will fail to do so for the other one. I said that the error can be misleading because user could see "NVML not loaded" in the log which is not the truth, it's just that NVML can't handle non–Nvidia GPUs.